Make Solidus Braintree compatible with Starter Frontend

gsmendoza opened the pull request:
## Summary

Closes https://github.com/solidusio/solidus_braintree/issues/91.

## Scope

This doesn't include updating the InstallGenerator to copy the frontend code to the user's app. That will be handled in #104 .

Likewise, updating InstallerGenerator to match the one from SolidusPaypalCommercePlatform will be done in #104.

## Demo

[screencast-bpconcjcammlapcogcnnelfmaeghhagj-2023.02.16-08_54_40.webm](https://user-images.githubusercontent.com/61476/219230996-30cdf101-7633-40a4-8048-bad7cf905be6.webm)

## Postponed tasks

* ~[ ] Test Apple Pay checkout - Need iOS device to test.~ - Can't do for now. See #105.
* ~[ ] Test Venmo checkout - Not successful in doing https://github.com/solidusio/solidus_braintree#testing in master (SolidusFrontend). See https://github.com/solidusio/solidus_braintree/pull/102#issuecomment-1436467805.~ - Postponed this to https://github.com/solidusio/solidus_braintree/issues/106.

## TODOs

### PR updates - first round

* [x] @elia - Document purpose for local OrderWalkthrough in commit message.
* [x] @elia - Install VCR and Webmock as development dependencies in client app.
* [x] @waiting-for-dev - Convert decorators to modules and include them to SSF controllers.
* [x] @waiting-for-dev - Test if the payment method is working without problems in the backend.
* [x] @kennyadsl - Remove CVV link.
* [x] Update existing payment partial based on recent updates to SSF.
  * [x] Fix: selecting an existing payment method returns an "All fields are empty" error
  * [x] Fix: Given I have chosen an existing payment method, I can still click on the Paypal button in the new payment method option.
  * [x] Fix spec.

### Fix CI

* [x] Fix transactions controller spec.
* [x] Update dummy app generation to fix system specs
* [x] Fix lint issues

### Production environment

* [x] Ensure that assets are precompiled 

### Development environment

* [x] Fix sandbox

### Documentation

* [x] Update README: no longer compatible with `solidus_frontend`.

### Testing

* [x] Test Credit card checkout with 3D secure.
* [x] Test PayPal checkout.
* [x] Test PayPal checkout with cart button.

### PR documentation

* [x] Demonstrate checkout.

## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] [I have used clear, explanatory commit messages](https://github.com/solidusio/.github/blob/main/CONTRIBUTING.md#writing-good-commit-messages).

The following are not always needed:

- 📖 I have updated the README to account for my changes.
- 📑 I have documented new code [with YARD](https://www.rubydoc.info/gems/yard/file/docs/Tags.md).
- 🛣️ I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).
- ✅ I have added automated tests to cover my changes.
- 📸 I have attached screenshots to demo visual changes.

elia commented on solidus_braintree.gemspec:
```diff
@@ -34,11 +34,11 @@ Gem::Specification.new do |spec|
   spec.add_dependency 'solidus_api', ['>= 2.4.0', '< 4']
   spec.add_dependency 'solidus_core', ['>= 2.4.0', '< 4']
   spec.add_dependency 'solidus_support', ['>= 0.8.1', '< 1']
+  spec.add_dependency 'vcr'
+  spec.add_dependency 'webmock'
```
If this is required in the target app as for running specs maybe it would be better to have them added to the gemfile by the installer, thoughts?

gsmendoza replied:
@elia That makes sense since they're only used for the specs and are not actual dependencies of the extension. I'll try letting the installer append them to the Gemfile :+1:

gsmendoza replied:
@elia I'm having some issues getting InstallGenerator to install the vcr and webmock gems in CI. See https://app.circleci.com/pipelines/github/solidusio/solidus_braintree/255/workflows/13c4ce91-4091-4965-b09e-d60a26cf9e27. I've moved back the gems to the gemspec for now. Will investigate tomorrow...

elia reviewed (commented):
LGTM overall, just left a couple of nits

gsmendoza commented on app/views/spree/shared/_braintree_hosted_fields.html.erb:
```diff
@@ -23,7 +23,7 @@
     <div class="input" id="card_code<%= payment_method.id %>"></div>
 
-    <a href="/content/cvv" class="info cvvLink" target="_blank">
+    <a href="/solidus_braintree/content/cvv" class="info cvvLink" target="_blank">
```
@elia @kennyadsl The CVV popup was removed from StarterFrontend in https://github.com/solidusio/solidus_starter_frontend/pull/10. However, SolidusBraintree has a link to the popup. To fix the issue, I copied the CVV page from SolidusFrontend. At the moment, it's rudimentary: the link just opens an unformatted page containing the CVV information

Moving forward, we have several options:

1. Restore the CVV popup in StarterFrontend.
2. Retain the CVV popup here and let each extension render its own popup.
3. Remove the CVV link and popup altogether.

Your thoughts?

gsmendoza replied:
Okay, let's remove the CVV link :+1:

gsmendoza commented on spec/features/backend/new_payment_spec.rb:
```diff
@@ -1,5 +1,4 @@
 require 'spec_helper'
-require 'spree/testing_support/order_walkthrough'
```
@elia I missed adding an explanation to the commit. Will add it :+1:

waiting-for-dev commented on app/decorators/controllers/solidus_braintree/carts_controller_decorator.rb:
```diff
@@ -1,11 +1,11 @@
     end
 
-    ::Spree::OrdersController.prepend(self) if SolidusSupport.frontend_available?
+    CartsController.prepend(self)
```
As the user owns `CartsController`, shouldn't this code be injected by the generator? Probably in a neat way through the inclusion of a module or whatever.

gsmendoza replied:
@waiting-for-dev Yes, I think it should be possible to include the decorators to the controllers. Will give it a try :+1:

waiting-for-dev commented on app/decorators/controllers/solidus_braintree/checkout_controller_decorator.rb:
```diff
@@ -6,6 +6,6 @@ def self.prepended(base)
     end
 
-    ::Spree::CheckoutController.prepend(self) if SolidusSupport.frontend_available?
+    ::CheckoutsController.prepend(self)
```
Same here. It probably doesn't make sense to monkey patch what is owned.

waiting-for-dev reviewed (commented):
Thanks, @gsmendoza. I left two comments about the need to monkey patch what the user owns. Also, did you try if the payment method is working without problems in the backend? See https://github.com/solidusio/solidus_stripe/issues/165 for more context.

gsmendoza commented:
@waiting-for-dev I am able to configure the SolidusBraintree payment method in the backend:

![image](https://user-images.githubusercontent.com/61476/219544442-dec578b5-d76b-4aa5-90b8-cf5d60169605.png)

waiting-for-dev reviewed (approved):
Thanks!

gsmendoza commented:
I tried to test the Venmo checkout in the *legacy* frontend by faking the nonce:

```diff
diff --git a/app/assets/javascripts/solidus_braintree/venmo_button.js b/app/assets/javascripts/solidus_braintree/venmo_button.js
index 1e54779..debb219 100644
--- a/app/assets/javascripts/solidus_braintree/venmo_button.js
+++ b/app/assets/javascripts/solidus_braintree/venmo_button.js
@@ -51,7 +51,8 @@ SolidusBraintree.VenmoButton.prototype.initializeCallback = function() {
 };
 
 SolidusBraintree.VenmoButton.prototype.initializeVenmoSession = function() {
-  this.tokenize();
+  var payload = { 'nonce' : 'fake-venmo-account-nonce' }
+  handleVenmoSuccess(payload);
 };
 
 SolidusBraintree.VenmoButton.prototype.tokenize = function() {
```

With this change, clicking the Venmo button in the Payment step allows me to skip the QR popup and go directly to the Confirm step:

![image](https://user-images.githubusercontent.com/61476/220037526-e1e0a161-4cc8-4c83-abf9-4df5dc61cd0b.png)

However, placing the order results in the following error:

> Merchant account does not support payment instrument. (91577) Payment instrument type is not accepted by this merchant account. (91517)

![image](https://user-images.githubusercontent.com/61476/220037609-a2de0282-220a-4864-93f1-adf4d0efcd6d.png)

I asked @RyanofWoods for help with this issue. He also tested it just now and got the same error.

Concerning the errors, https://support.chargebee.com/support/solutions/articles/220172-error-91517-91577-payment-instrument-is-not-accepted-by-this-merchant-account-merchant-account-does- says

> 91517 and 91577 are validation errors. These errors indicate that a particular card type (American Express for example) or a token ([read about tokens here](https://en.wikipedia.org/wiki/Tokenization_(data_security))) associated with a particular card type isn’t accepted by your merchant account. To resolve, you would need to reach out to your merchant account provider and enable the card type in question.

gsmendoza commented:
Regarding https://github.com/solidusio/solidus_braintree/pull/102#issuecomment-1436467805, I also tried changing my location to a US server via a VPN, but I'm still getting the "Merchant account does not support payment instrument." error.

gsmendoza commented:
I've updated the PR based on your feedback and recent changes. Please see the following commits for the changes:

* @elia - Replace OrderWalkthrough with local version - Updated commit message
* @elia - Install vcr and webmock as test gems
* @waiting-for-dev - Inject SolidusBraintree dependencies into controllers
* @kennyadsl - Remove CVV link
* @elia - I added the following commits in response to https://github.com/solidusio/solidus_starter_frontend/pull/321:
  - Remove radio button from existing Braintree payment partial
  - Tokenize hosted fieldset Braintree form only if fieldset is enabled
  - Fix payment method container selector in spec
  - Do not allow buttons in disabled payment-step to be clickable
  
@elia, regarding "Do not allow buttons in disabled payment-step to be clickable", it might be possible to disable the PayPal, Venmo, and Apple Pay buttons via JS. However, we'll still need to discuss whether to implement that in StarterFrontend or within each payment extension. For the meantime, the CSS workaround prevents the buttons from being clicked.

Here's a quick demo of the payment page:

[2023-02-22-payment.webm](https://user-images.githubusercontent.com/61476/220540919-6bfc6f30-ea90-4f37-a99f-c8a73c72ad45.webm)

gsmendoza commented:
@elia @waiting-for-dev Merging this now so that we can move forward with the https://github.com/solidusio/solidus_braintree/issues/105 and https://github.com/solidusio/solidus_braintree/issues/106 tickets. If you have additional feedback, I can work on them while working on https://github.com/solidusio/solidus_braintree/issues/104.
